### PR TITLE
Relax Arrow pin

### PR DESCRIPTION
**Pull request by vyasr**
## Description

With this change, cudf users can install any version of pyarrow greater than 14. This is the minimum version supporting the C Data Interface, which is a requirement for us (it may be possible to relax in principle, but would require changes to the cudf/pylibcudf code). A few tests are skipped due to bugs or missing features in older versions of pyarrow.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by galipremsagar — CHANGES_REQUESTED** [review]
In my local testing I found that the minimum required version might be 15.0.0 because `__arrow_c_stream__` seems to be added in 15.0.0: 

https://github.com/apache/arrow/issues/38717

But it shows up in changelog of 16.0.0 only(and not 15.0.0):

Screenshot 2024-08-28 at 11 46 11 AM

However, if I install a 14.0.0 version of pyarrow, I still have `__arrow_c_stream__` - It might have been possibly back-ported. is that the case? I'm not able to find history of this being back-ported.

These changes look good. But should be rather keep an upper bound pin on the `pyarrow` version because of the same reasons we do for other libraries?

Locally verified `cudf` & `dask-cudf` work quite well for all of the following versions:

- [x] 14.0.0
- [x] 15.0.0
- [x] 15.0.1
- [x] 16.0.0
- [x] 16.1.0
- [x] 17.0.0

**Comment by vyasr** [comment]
The original capsule interface was added in https://github.com/apache/arrow/pull/37797/. https://github.com/apache/arrow/pull/39455 looks to be specifically providing the stream capsule for chunked arrays. I believe that we use the `__arrow_c_array__` protocol for those instead and rely on combining chunks, hence why Arrow 14 seems to be sufficient. Let me double check what's happening on 14 to be sure.

**Comment by vyasr** [comment]
Well, the answer is not quite as good as I had hoped. I was right that our current conversion function for ChunkedArray tries to use the `__arrow_c_array__` attribute, but that doesn't actually exist so it will just error out. I do appear to be correct that we are always combining chunks as needed before attempting such a conversion though because I was able to run the entire test suite with pylibcudf modified to raise an exception if we ever pass a `ChunkedArray` to `from_arrow`. Therefore, I propose that we leave the pinning on this PR as is (allow arrow 14), but remove the erroneous overload of `from_arrow` for ChunkedArray. If we choose to implement it correctly in the future, we can either bump the minimum pyarrow version or gate just that bit of functionality on the right attributes existing. Does that sound OK?

**Comment by galipremsagar** [comment]
> Well, the answer is not quite as good as I had hoped. I was right that our current conversion function for ChunkedArray tries to use the `__arrow_c_array__` attribute, but that doesn't actually exist so it will just error out. I do appear to be correct that we are always combining chunks as needed before attempting such a conversion though because I was able to run the entire test suite with pylibcudf modified to raise an exception if we ever pass a `ChunkedArray` to `from_arrow`. Therefore, I propose that we leave the pinning on this PR as is (allow arrow 14), but remove the erroneous overload of `from_arrow` for ChunkedArray. If we choose to implement it correctly in the future, we can either bump the minimum pyarrow version or gate just that bit of functionality on the right attributes existing. Does that sound OK?

Yup, sounds good to me. 👍

**Comment by vyasr** [comment]
Cool. Added upper bounds as well.

**Review by galipremsagar — APPROVED** [review]
This looks good to go! Thanks @vyasr !

File: dependencies.yaml
Comment (jameslamb) [inline dependencies.yaml:145]: I still see a runtime dependency on `pyarrow` in `libcudf`'s `pyproject.toml` on your branch: https://github.com/vyasr/cudf/blob/228106a599aed68b8d35eca7498408022552194c/python/libcudf/pyproject.toml#L40-L42

I think you'll need to manually remove that... `rapids-dependency-file-generator` won't automatically recognize that it's no longer being asked to manage a section that it had previously populated.

**Review by jameslamb — APPROVED** [review]
Just one comment, but marking this "approve" so you don't need to wait on another review from me after addressing that. @galipremsagar 's testing in https://github.com/rapidsai/cudf/pull/16681#pullrequestreview-2266781557 and the discussion after that makes me very confident in this change.

File: dependencies.yaml
Comment (vyasr) [reply dependencies.yaml:145]: Great catch! It didn't occur to me that the libcudf pyproject.toml had a runtime list with just a single dependency.

**Comment by vyasr** [comment]
For the record in case it helps inspire future confidence, independently of Prem I also ran the test suite on all the Python versions that he listed with similar results (which is what revealed the skips that I added to the test suite).

**Pull request merged**
